File: lint-arwaky/src-python/infrastructure/os_fs_scanner.py
"""Implementation of IFileSystemScanner using the standard os module."""

from __future__ import annotations

from ..taxonomy import (
    Count,
    FileContentVO,
    FilePath,
    PatternList,
    SuccessStatus,
    BooleanVO,
    Identity,
    Timestamp,
    FileSystemError,
    ErrorMessage,
)

import logging
import os
from typing import Iterator

from ..contract import IFileSystemPort

logger = logging.getLogger(__name__)

try:
    from auto_linter import auto_linter_rust
    RUST_AVAILABLE = True
except ImportError:
    RUST_AVAILABLE = False


class OSFileSystemAdapter(IFileSystemPort):
    def __init__(self) -> None:
        self._exists_cache: dict[FilePath, BooleanVO] = {}
        self._is_file_cache: dict[FilePath, BooleanVO] = {}
        self._is_dir_cache: dict[FilePath, BooleanVO] = {}
        self._line_count_cache: dict[tuple[FilePath, Timestamp], Count] = {}

    def walk(
        self, path: FilePath, ignored_patterns: PatternList | None = None
    ) -> Iterator[FilePath]:
        path_str = str(path)
        ignored = ignored_patterns.values if ignored_patterns else []

        if os.path.isfile(path_str):
            yield path
            return

        if RUST_AVAILABLE:
            try:
                paths = auto_linter_rust.walk_files(path_str, ignored)
                for p in paths:
                    yield FilePath(value=p)
                return
            except Exception as e:
                logger.warning(f"Rust fs scanner failed: {e}. Falling back to Python os.walk.")

        for root, dirs, files in os.walk(path_str):
            # Prune ignored directories
            dirs[:] = [d for d in dirs if d not in ignored]

            for f in files:
                full_path = os.path.join(root, f)
                # Check if any part of the relative path is ignored
                rel_path = os.path.relpath(full_path, path_str)
                if any(part in ignored for part in rel_path.split(os.sep)):
                    continue
                yield FilePath(value=full_path)

    def is_directory(self, path: FilePath) -> SuccessStatus:
        if path not in self._is_dir_cache:
            self._is_dir_cache[path] = BooleanVO(value=os.path.isdir(str(path)))
        return SuccessStatus(value=self._is_dir_cache[path])

    def is_file(self, path: FilePath) -> SuccessStatus:
        if path not in self._is_file_cache:
            self._is_file_cache[path] = BooleanVO(value=os.path.isfile(str(path)))
        return SuccessStatus(value=self._is_file_cache[path])

    def get_relative_path(self, path: FilePath, start: FilePath) -> FilePath:
        return FilePath(value=os.path.relpath(str(path), str(start)).replace("\\", "/"))

    def read_text(self, path: FilePath) -> FileContentVO | FileSystemError:
        try:
            with open(str(path), "r", encoding="utf-8") as f:
                return FileContentVO(value=f.read())
        except Exception as e:
            logger.warning("Failed to read text from %s: %s", path, e, exc_info=True)
            return FileSystemError(
                path=path,
                message=ErrorMessage(value=f"Failed to read file: {e}")
            )

    def get_line_count(self, path: FilePath) -> Count:
        try:
            mtime = os.path.getmtime(str(path))
            timestamp = Timestamp(value=str(mtime))
            cache_key = (path, timestamp)
            if cache_key in self._line_count_cache:
                return self._line_count_cache[cache_key]

            with open(str(path), "r", encoding="utf-8") as f:
                count_val = sum(1 for _ in f)
                count = Count(value=count_val)
                self._line_count_cache[cache_key] = count
                return count
        except Exception:
            logger.warning("Failed to count lines in %s", path, exc_info=True)
            return Count(value=0)

    def exists(self, path: FilePath) -> SuccessStatus:
        if path not in self._exists_cache:
            self._exists_cache[path] = BooleanVO(value=os.path.exists(str(path)))
        return SuccessStatus(value=self._exists_cache[path])

    def get_parent(self, path: FilePath) -> FilePath:
        parent = os.path.dirname(str(path)) or "."
        return FilePath(value=parent)

    def invalidate_cache(self, path: FilePath | None = None) -> None:
        """Clear cache for a specific path, or all paths if path is None."""
        if path is None:
            self._exists_cache.clear()
            self._is_file_cache.clear()
            self._is_dir_cache.clear()
            self._line_count_cache.clear()
        else:
            self._exists_cache.pop(path, None)
            self._is_file_cache.pop(path, None)
            self._is_dir_cache.pop(path, None)
            keys_to_remove = [k for k in self._line_count_cache if k[0] == path]
            for k in keys_to_remove:
                self._line_count_cache.pop(k, None)

    def write_text(
        self, path: FilePath, content: FileContentVO, mode: Identity | None = None
    ) -> SuccessStatus | FileSystemError:
        try:
            mode_str = str(mode) if mode else "w"
            py_mode = "a" if mode_str in ("a", "append") else "w"
            with open(str(path), py_mode, encoding="utf-8") as f:
                f.write(str(content))
            self.invalidate_cache(path)
            return SuccessStatus(value=BooleanVO(value=True))
        except Exception as e:
            logger.warning("Failed to write text to %s: %s", path, e, exc_info=True)
            return FileSystemError(
                path=path,
                message=ErrorMessage(value=f"Failed to write file: {e}")
            )

    def glob(self, pattern: Identity) -> Iterator[FilePath]:
        import glob

        for p in glob.glob(str(pattern), recursive=True):
            yield FilePath(value=p)

    def get_cwd(self) -> FilePath:
        return FilePath(value=os.getcwd().replace("\\", "/"))

    def get_basename(self, path: FilePath) -> Identity:
        return Identity(value=os.path.basename(str(path)))

    def path_join(self, *parts: Identity) -> FilePath:
        return FilePath(value=os.path.join(*[str(p) for p in parts]).replace("\\", "/"))

    def read_file(self, path: FilePath) -> FileContentVO:
        return self.read_text(path)

File: lint-arwaky/src-rust/infrastructure/os_fs_scanner.rs
/// Implementation of IFileSystemPort using standard std::fs.
use crate::contract::IFileSystemPort;
use crate::taxonomy::*;
use async_trait::async_trait;
use std::fs;
use std::path::{Path, PathBuf};

pub struct OSFileSystemAdapter;

impl OSFileSystemAdapter {
    pub fn new() -> Self {
        Self
    }

    fn walk_recursive(&self, dir: &Path, ignored: &[String], results: &mut Vec<FilePath>) {
        if dir.is_file() {
            results.push(FilePath::new(dir.to_string_lossy().to_string()));
            return;
        }

        if let Ok(entries) = fs::read_dir(dir) {
            for entry in entries.flatten() {
                let path = entry.path();
                let name = path.file_name().and_then(|n| n.to_str()).unwrap_or("");

                if ignored.contains(&name.to_string()) {
                    continue;
                }

                if path.is_dir() {
                    self.walk_recursive(&path, ignored, results);
                } else {
                    results.push(FilePath::new(path.to_string_lossy().to_string()));
                }
            }
        }
    }
}

#[async_trait]
impl IFileSystemPort for OSFileSystemAdapter {
    async fn walk(&self, path: &FilePath, ignored_patterns: Option<&PatternList>) -> Vec<FilePath> {
        let root = Path::new(&path.value);
        let ignored = ignored_patterns
            .map(|p| p.values.clone())
            .unwrap_or_default();
        let mut results = Vec::new();
        self.walk_recursive(root, &ignored, &mut results);
        results
    }

    async fn is_directory(&self, path: &FilePath) -> SuccessStatus {
        SuccessStatus::new(BooleanVO::new(Path::new(&path.value).is_dir()))
    }

    async fn is_file(&self, path: &FilePath) -> SuccessStatus {
        SuccessStatus::new(BooleanVO::new(Path::new(&path.value).is_file()))
    }

    async fn get_relative_path(&self, path: &FilePath, start: &FilePath) -> FilePath {
        let p = Path::new(&path.value);
        let s = Path::new(&start.value);
        match p.strip_prefix(s) {
            Ok(rel) => FilePath::new(rel.to_string_lossy().to_string()),
            Err(_) => path.clone(),
        }
    }

    async fn read_text(&self, path: &FilePath) -> Result<FileContentVO, FileSystemError> {
        self.read_file(path).await
    }

    async fn get_line_count(&self, path: &FilePath) -> Count {
        if let Ok(content) = fs::read_to_string(&path.value) {
            Count::new(content.lines().count() as i64)
        } else {
            Count::new(0)
        }
    }

    async fn exists(&self, path: &FilePath) -> SuccessStatus {
        SuccessStatus::new(BooleanVO::new(Path::new(&path.value).exists()))
    }

    async fn get_parent(&self, path: &FilePath) -> FilePath {
        let p = Path::new(&path.value);
        match p.parent() {
            Some(parent) => FilePath::new(parent.to_string_lossy().to_string()),
            None => path.clone(),
        }
    }

    async fn write_text(
        &self,
        path: &FilePath,
        content: &FileContentVO,
        _mode: Option<&Identity>,
    ) -> Result<SuccessStatus, FileSystemError> {
        match fs::write(&path.value, &content.value) {
            Ok(_) => Ok(SuccessStatus::new(BooleanVO::new(true))),
            Err(e) => Err(FileSystemError {
                path: path.clone(),
                message: ErrorMessage::new(e.to_string()),
                ..Default::default()
            }),
        }
    }

    async fn glob(&self, pattern: &Identity) -> Vec<FilePath> {
        // Simple mock glob
        vec![]
    }

    async fn get_cwd(&self) -> FilePath {
        let cwd = std::env::current_dir().unwrap_or_else(|_| PathBuf::from("."));
        FilePath::new(cwd.to_string_lossy().to_string())
    }

    async fn get_basename(&self, path: &FilePath) -> Identity {
        let p = Path::new(&path.value);
        Identity::new(
            p.file_name()
                .and_then(|n| n.to_str())
                .unwrap_or("")
                .to_string(),
        )
    }

    async fn path_join(&self, parts: &[Identity]) -> FilePath {
        let mut path = PathBuf::new();
        for part in parts {
            path.push(&part.value);
        }
        FilePath::new(path.to_string_lossy().to_string())
    }

    async fn read_file(&self, path: &FilePath) -> Result<FileContentVO, FileSystemError> {
        match fs::read_to_string(&path.value) {
            Ok(content) => Ok(FileContentVO::new(content)),
            Err(e) => Err(FileSystemError {
                path: path.clone(),
                message: ErrorMessage::new(e.to_string()),
                ..Default::default()
            }),
        }
    }
}
